servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.SingleSource.Subscriber;
+import io.servicetalk.concurrent.internal.DelayedCancellable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+final class AmbSingles<T> extends Single<T> {
+    private final Single<? extends T>[] singles;
+
+    @SafeVarargs
+    AmbSingles(final Single<? extends T>... singles) {
+        for (Single<? extends T> single : singles) {
+            requireNonNull(single);
+        }
+        this.singles = singles;
+    }
+
+    AmbSingles(final Iterable<Single<? extends T>> singles) {
+        List<Single<? extends T>> allSingles = new ArrayList<>();
+        for (Single<? extends T> single : singles) {
+            allSingles.add(requireNonNull(single));
+        }
+        @SuppressWarnings({"unchecked", "SuspiciousToArrayCall"})
+        Single<? extends T>[] singlesArr = (Single<? extends T>[]) allSingles.toArray(new Single[0]);
+        this.singles = singlesArr;
+    }
+
+    @Override
+    protected void handleSubscribe(final Subscriber<? super T> subscriber) {
+        final Cancellable[] cancellables = new Cancellable[singles.length];
+        final State<T> state = new State<>(subscriber);
+        subscriber.onSubscribe(state);
+        try {
+            for (int i = 0; i < singles.length; i++) {
+                AmbSubscriber<T> sub = new AmbSubscriber<>(state);
+                cancellables[i] = sub;
+                singles[i].subscribeInternal(sub);
+            }
+        } catch (Throwable t) {
+            state.delayedCancellable(CompositeCancellable.create(cancellables));
+            state.tryError(t);
+            return;
+        }
+        state.delayedCancellable(CompositeCancellable.create(cancellables));
+    }
+
+    static final class AmbSubscriber<T> extends DelayedCancellable implements Subscriber<T> {
```

Comment:
In this case we are using `DelayedCancellable` due to the delay nature (other `Subscriber` can terminate before this gets `onSubscribe()`) and not for concurrent access protection. Did you think otherwise?

Reply:
The `Cancellable` maybe invoked concurrently from multiple threads (e.g. `singles[i].subscribeInternal(sub);` and `state.delayedCancellable(CompositeCancellable.create(cancellables));`). So in addition to "delaying" semantics the assumption is it will also provide "invoke at most once in a thread safe manner" semantics. Based upon our usage they are closely related , I'm just suggesting clarifying this in the API contract to make this more explicit.

Reply:
Ok ya I understand your suggestion. 

The delay nature essentially introduces potential concurrent invocation as the reason why it is delayed is because we want to use it before the actual `Cancellable` arrives. It isn't unique about this usage. Let me improve the documentation in a follow up.